jep/0000/README.adoc

```diff
@@ -0,0 +1,225 @@
+
+From the usability point of view, currently a Jenkins installation sizing has to have into account the expected amount of disk needed or a restart, disk reprovisioning, disk swap,... instead of a infinite scale, pay per use that a blob store can offer. Cloud installations can automatically configure the blob store to use offering a much better first time user experience.
+
+Several alternatives exist today but they all require changes to all the pipelines and job definitions to explicitly choose the backend to send artifacts to.
+
+* https://plugins.jenkins.io/s3[S3 plugin]
+* https://github.com/jenkinsci/pipeline-aws-plugin[pipeline-aws-plugin]
+* https://github.com/jenkinsci/windows-azure-storage-plugin[windows-azure-storage-plugin]
+* https://github.com/jenkinsci/google-storage-plugin[google-storage-plugin]
+
+Even without this JEP people can use things like the S3 plugin to upload and download artifacts.
+But since the use of S3, and details about location, are baked into the script, we cannot publish general examples like https://jenkins.io/doc/pipeline/tour/tests-and-artifacts/[tests-and-artifacts] that are actually ready for people to use.
+That would contradict one of the goals of Essentials, that you can get a reasonable workflow going in a few minutes.
+
+Without `ArtifactManager` and `VirtualFile` integration, a number of integrations between plugins are impossible.
+For example, using only the S3 plugin, if you wish to copy artifacts from an upstream build, you cannot use the Copy Artifact plugin; you would need to devise your own system for passing an S3 bucket/path from the upstream build to the downstream build.
+When https://issues.jenkins-ci.org/browse/JENKINS-45455[JENKINS-45455] is implemented, unstash from S3 will work automatically in a restarted Pipeline build to copy files stashed by the original build.
+Using only the S3 plugin, you would need to think about saving bucket/path to a variable that could be read by the restarted build.
+Blue Ocean will display an Artifacts tab for files uploaded to S3 via `archiveArtifacts`; with only the S3 plugin, you would need to go to Classic UI.
+
+Core APIs already existed for customized artifact storage, but lacked the crucial capability to offer pre-signed URLs, making it impossible to provide a satisfactory S3 implementation.
+Only customized master-side storage (such as with Compress Artifacts) was really practical.
+
+== Reasoning
+
+AWS is the focus as it is the most widely used cloud provider, S3 being the prevalent blob store.
+Equivalent features to S3 exist in other cloud providers and artifact repositories.
+
+The S3 implementation also uses http://jclouds.apache.org[Apache JClouds] that abstracts most of the implementation from the underlying blob store.
+
+== Backwards Compatibility
+
+Existing plugins using `ArtifactManager` API will continue to work using the new selected implementation.
```

Comment:
What about plugins NOT using this API? Will they continue working if this new Artifact Manager is disabled?

Would be also nice to provide a list of known plugins which work with artifacts directly without using this API (and hence will break in new impl). At the same point mention that the current Filesystem-based storage will continue working as before for that plugins